main_test.go

```diff
@@ -0,0 +1,434 @@
+package main
+
+import (
+	"context"
+	"os"
+	"path"
+	"testing"
+	"time"
+
+	"github.com/efficientgo/tools/core/pkg/testutil"
+	"github.com/go-kit/log"
+	"github.com/observatorium/obsctl/pkg/config"
+	monitoringv1 "github.com/prometheus-operator/prometheus-operator/pkg/apis/monitoring/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+)
+
+type testRulesLoader struct{}
+
+func (r *testRulesLoader) GetPrometheusRules() ([]*monitoringv1.PrometheusRule, error) {
+	return nil, nil
+}
+
+func (r *testRulesLoader) GetTenantRuleGroups(_ []*monitoringv1.PrometheusRule) map[string]monitoringv1.PrometheusRuleSpec {
+	return map[string]monitoringv1.PrometheusRuleSpec{
+		"test": {},
+	}
+}
+
+type testRulesSyncer struct {
+	counter int
+}
+
+func (r *testRulesSyncer) InitOrReloadObsctlConfig() error {
+	return nil
+}
+
+func (r *testRulesSyncer) SetCurrentTenant(tenant string) error {
+	return nil
+}
+
+func (r *testRulesSyncer) ObsctlMetricsSet(rules monitoringv1.PrometheusRuleSpec) error {
+	r.counter++
+	return nil
+}
+
+func TestSyncLoop(t *testing.T) {
+	rl := &testRulesLoader{}
+	rs := &testRulesSyncer{}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	time.AfterFunc(25*time.Second, func() { cancel() })
+
+	SyncLoop(ctx, log.NewLogfmtLogger(log.NewSyncWriter(os.Stderr)), rl, rs, 5)
+
+	// <-ctx.Done()
+	testutil.Equals(t, 4, rs.counter)
+}
+
+func TestInitOrReloadObsctlConfig(t *testing.T) {
+	o := &obsctlRulesSyncer{ctx: context.TODO(), logger: log.NewLogfmtLogger(log.NewSyncWriter(os.Stderr)), skipClientCheck: true}
+	testutil.Ok(t, os.Setenv("OBSERVATORIUM_URL", "http://yolo.com/"))
+	testutil.Ok(t, os.Setenv("OIDC_AUDIENCE", "test"))
+	testutil.Ok(t, os.Setenv("OIDC_CLIENT_ID", "test"))
+	testutil.Ok(t, os.Setenv("OIDC_CLIENT_SECRET", "test"))
+	testutil.Ok(t, os.Setenv("OIDC_ISSUER_URL", "http://yolo-auth.com"))
+
+	testOIDC := &config.OIDCConfig{
+		Audience:     "test",
+		ClientID:     "test",
+		ClientSecret: "test",
+		IssuerURL:    "http://yolo-auth.com",
+	}
+
+	for _, tc := range []struct {
+		name             string
+		tenants          string
+		prexistingConfig *config.Config
+		wantConfig       *config.Config
+	}{
+		{
+			name: "already existing config with one tenant",
```

Comment:
I wonder if it makes sense to test with an empty config as well?

Reply:
Will add a test case!